coderd/workspaceagents.go

```diff
@@ -1644,34 +1661,26 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
 			refreshTicker.Reset(refreshInterval)
 		}
 
-		_ = sendEvent(ctx, codersdk.ServerSentEvent{
+		_ = sseSendEvent(ctx, codersdk.ServerSentEvent{
 			Type: codersdk.ServerSentEventTypeData,
-			Data: convertWorkspaceAgentMetadata(lastDBMeta),
+			Data: convertWorkspaceAgentMetadata(lastMetadata),
 		})
 	}
 
-	// Note: we previously used a debounce here, but when the rate of metadata updates was too
-	// high the debounce would never fire.
-	//
-	// The rate-limit has its own caveat. If the agent sends a burst of metadata
-	// but then goes quiet, we will never pull the new metadata and the frontend
-	// will go stale until refresh. This would only happen if the agent was
-	// under extreme load. Under normal operations, the interval between metadata
-	// updates is constant so there is no burst phenomenon.
-	pubsubRatelimit := rate.NewLimiter(rate.Every(time.Second), 2)
-	if flag.Lookup("test.v") != nil {
-		// We essentially disable the rate-limit in tests for determinism.
-		pubsubRatelimit = rate.NewLimiter(rate.Every(time.Second*100), 100)
-	}
+	// Note: we tried using a ratelimit and debounce here before but it was
+	// pretty buggy.
+	pubsubUpdates := make(chan database.UpdateWorkspaceAgentMetadataParams, 8)
 
 	// Send metadata on updates, we must ensure subscription before sending
 	// initial metadata to guarantee that events in-between are not missed.
-	cancelSub, err := api.Pubsub.Subscribe(watchWorkspaceAgentMetadataChannel(workspaceAgent.ID), func(_ context.Context, _ []byte) {
-		allow := pubsubRatelimit.Allow()
-		log.Debug(ctx, "received metadata update", "allow", allow)
-		if allow {
-			sendMetadata(true)
+	cancelSub, err := api.Pubsub.Subscribe(watchWorkspaceAgentMetadataChannel(workspaceAgent.ID), func(_ context.Context, byt []byte) {
+		var update database.UpdateWorkspaceAgentMetadataParams
+		err = unmarshalGob(byt, &update)
+		if err != nil {
+			api.Logger.Error(ctx, "failed to unmarshal pubsub message", slog.Error(err))
+			return
 		}
+		pubsubUpdates <- update
```

Comment:
If `sendMetadata` takes a long time, there's a chance we'll fill up the 8 buffer slots we've got here and then hang.

We could instead append to a slice here and use a buffer of 1 instead to signal that there's new data in the slice to process.

Reply:
I think we want to cap the buffer size at something though, at which point we're back to the same problem.

Reply:
Do you mean preventing the slice from growing unboundedly? I suppose we can just limit it's max size and discard the oldest once it has more than X entries? As for notifying we can just do a select with a default on a 1 size chan.

Reply:
hmm I can actually just use a map on the key